From a3aa00b38232bc94d6a99f03bd415b8b06d5525f Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Wed, 24 Aug 2016 16:26:22 +0800
Subject: [PATCH 0435/1345] fix: mvneta: remove 1000BaseT_Half from auto-nego
 supported mode

commit  37aa991c5b640c90e7bd626119246e52d46ac46d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- It fix issue SYSTEMSW-2644 that Neta SoCs should not support
  1000Base Half mode, such as A38x, A3700, A370 and AXP
- This patch add valdation on advertising mode to prevent from
  the misconfigurations.

Change-Id: Ic8039e16992d05a62284927a65ce214f6d1dc327
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/32149
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvneta.c |   27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 7d294dc..e360026 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -3452,7 +3452,8 @@ static int mvneta_mdio_probe(struct mvneta_port *pp)
 	phy_ethtool_get_wol(phy_dev, &wol);
 	device_set_wakeup_capable(&pp->dev->dev, !!wol.supported);
 
-	phy_dev->supported &= PHY_GBIT_FEATURES;
+	/* Neta does not support 1000baseT_Half */
+	phy_dev->supported &= (PHY_GBIT_FEATURES & (~SUPPORTED_1000baseT_Half));
 	phy_dev->advertising = phy_dev->supported;
 
 	pp->link    = 0;
@@ -3753,6 +3754,27 @@ static int mvneta_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 
 /* Ethtool methods */
 
+/* Check speed and duplex when set auto-nego with ethtool */
+static int mvneta_spd_dplx_valid(struct mvneta_port *pp,
+				 struct ethtool_cmd *cmd)
+{
+	int ret = 0;
+	u32 speed = ethtool_cmd_speed(cmd);
+
+	if ((speed + cmd->duplex) == (SPEED_1000 + DUPLEX_HALF)) {
+		/* When auto-nego disabled, 1000Base-Half is illegal.
+		 * When auto-nego enabled, 1000Base-Half is invalid,
+		 * but no error return for this, ethtool will show results.
+		 */
+		if (cmd->autoneg == AUTONEG_DISABLE) {
+			netdev_err(pp->dev, "Unsupported Speed/Duplex configuration\n");
+			ret = -EINVAL;
+		}
+	}
+
+	return ret;
+}
+
 /* Set link ksettings (phy address, speed) for ethtools */
 static int
 mvneta_ethtool_set_link_ksettings(struct net_device *ndev,
@@ -3764,6 +3786,9 @@ static int mvneta_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 	if (!phydev)
 		return -ENODEV;
 
+	if (mvneta_spd_dplx_valid(pp, cmd))
+		return -EINVAL;
+
 	if ((cmd->base.autoneg == AUTONEG_ENABLE) != pp->use_inband_status) {
 		u32 val;
 
-- 
1.7.9.5

